### PR TITLE
Implement removal of orphan nodes

### PR DESCRIPTION
**Pull request by roysc**
Refactors the storage of values into a separate KV store, indexed by `hash(value)+key`. As noted in #14 loses the benefit of deduplication, while enabling pruning of records.

Completes work from https://github.com/celestiaorg/smt/pull/24.
Revises work from https://github.com/celestiaorg/smt/pull/36.

### CODE REVIEW

File: smt.go [hunk 1]
```diff
@@ -265,54 +270,72 @@ func (smt *SparseMerkleTree) updateWithSideNodes(path []byte, value []byte, side
...
+			return nil, err
+		}
+	}
+	if smt.prune {
+		// All remaining path nodes are orphaned
+		for i := 1; i < len(pathNodes); i++ {
+			if err := smt.nodes.Delete(pathNodes[i]); err != nil {
+				return nil, err
+			}
+		}
```
[the remark this answers is not in the document]
Comment (adlerjohn) [reply smt.go:309]: Ah it starts at `1`, nvm.

File: deepsubtree.go [hunk 2]
```diff
@@ -32,14 +27,20 @@ func NewDeepSparseMerkleSubTree(ms MapStore, hasher hash.Hash, root []byte) *Dee
...
 // an updatable proof should be used. See SparseMerkleTree.ProveUpdatable.
 func (dsmst *DeepSparseMerkleSubTree) AddBranch(proof SparseMerkleProof, key []byte, value []byte) error {
-	result, updates := verifyProofWithUpdates(proof, dsmst.Root(), key, value, dsmst.th.hasher)
+	result, updates, valueHash := verifyProofWithUpdates(proof, dsmst.Root(), key, value, dsmst.th.hasher)
 	if !result {
 		return ErrBadProof
 	}
 
+	if valueHash != nil {
+		if err := dsmst.values.Set(valueKey(key, valueHash), value); err != nil {
```
[the remark this answers is not in the document]
Comment (roysc) [reply deepsubtree.go:36]: With just `key/path -> value`, there is no way to retrieve values for past versions. I considered using `root+key` as @tzdybal [suggested](https://github.com/celestiaorg/smt/issues/14#issuecomment-804149760), but then lookup requires knowing the root at the value's last update. Seems we would need some sort of copy-on-write versioned backing store to make this efficient.

note - this is just `key+hash(value) -> value`, no extra hash.

File: smt.go [hunk 3]
```diff
@@ -155,22 +157,22 @@ func (smt *SparseMerkleTree) Delete(key []byte) ([]byte, error) {
...
+	sideNodes, pathNodes, oldLeafData, _, err := smt.sideNodesForRoot(path, root, false)
 	if err != nil {
 		return nil, err
 	}
 
 	var newRoot []byte
 	if bytes.Equal(value, defaultValue) {
 		// Delete operation.
-		newRoot, err = smt.deleteWithSideNodes(path, sideNodes, oldLeafHash, oldLeafData)
+		newRoot, err = smt.deleteWithSideNodes(key, path, sideNodes, pathNodes, oldLeafData)
```
[the remark this answers is not in the document]
Comment (roysc) [reply smt.go:168]: Ok, key is used for value deletion but I can pull that up

File: deepsubtree.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (adlerjohn) [reply deepsubtree.go:36]: Ah, that comment is outdated. Snapshotting and retrieving past version is outside the scope of this library. It can be accomplished at the db level. See https://github.com/cosmos/cosmos-sdk/blob/6a5a2de798d4e61302783ae84ecfe2958c7088d5/docs/architecture/adr-040-storage-and-smt-state-commitments.md.

> Seems we would need some sort of copy-on-write versioned backing store to make this efficient.

Potentially yes, and that's an interesting direction, but as above it's outside the scope of this library. The _caller_ can accomplish this by versioning the mapstores they pass in on SMT initialization.

TL;DR: use `path -> value`.

[the remark this answers is not in the document]
Comment (roysc) [reply deepsubtree.go:36]: >  The _caller_ can accomplish this by versioning the mapstores they pass in on SMT initialization.

This work is actually in support of that ADR - we'll be doing exactly that, so I'm basically indifferent to the indexing used here. I just didn't want to introduce breakage.

If just `path` is used, supporting past value access will be dropped, but I assume proofs for non-pruned roots should still be supported? Otherwise pruning effectively becomes the only behavior.

[the remark this answers is not in the document]
Comment (adlerjohn) [reply deepsubtree.go:36]: > If just path is used, supporting past value access will be dropped, but I assume proofs for non-pruned roots should still be supported? Otherwise pruning effectively becomes the only behavior.

Good point. Proofs against past versions of the tree isn't needed in this library, since that can be accomplished caller-side. So, pruning should be the only behavior (making the `prune` flag unnecessary).

[the remark this answers is not in the document]
Comment (musalbas) [reply deepsubtree.go:36]: Proofs against old roots is a test case specified in the current version of this library. If this feature is to be removed because it's considered out of scope, we should be sure that this can easily be replicated caller-side with a KV store with snapshot support. What KV store supports this exactly? I checked RocksDB, but its snapshot feature is basically copying the entire DB directory, which clearly isn't efficient.

[the remark this answers is not in the document]
Comment (adlerjohn) [reply deepsubtree.go:36]: The [ADR](https://github.com/cosmos/cosmos-sdk/blob/81747b4a9ea33f8823b04478a7cbf772d2102d97/docs/architecture/adr-040-storage-and-smt-state-commitments.md) states that the state commitment (i.e. the SMT, i.e. this library) isn't responsible for snapshots and historical queries, but rather the state storage is. Note that the state storage won't be a raw KV store, but will be a wrapper over one or more raw KV stores with additional logic.

> State Storage requirements:
> 
> - range queries
> - quick (key, value) access
> - creating a snapshot
> - historical versioning
> - pruning (garbage collection)
> 
> State Commitment requirements:
> 
> - fast updates
> - tree path should be short
> - pruning (garbage collection)

[the remark this answers is not in the document]
Comment (musalbas) [reply deepsubtree.go:36]: While there has been some benchmarking done on Rocks DB etc snapshots for the ADR, this isn't a widely supported feature by KV stores, and it's yet to be seen how well this will work in practice. RocksDB snapshots seem to be designed with backups use case in mind; [creating a new directory](https://rocksdb.org/blog/2015/11/10/use-checkpoints-for-efficient-snapshots.html) for every block could exhaust the inode limit on filesystems, given a chain with millions of blocks. BadgerDB seems to be more suitable as it's a key-value-version store, but this seems to be the only KV that supports this.

This library should act a general-purpose library. If someone wants to use this library for testing or research for example, they shouldn't have to setup extra scaffolding - or be tied to BadgerDB (the only KV store that seems to support efficient versioning) -  to get the basic feature of accessing previous roots.

For flexibility, this library should ideally support the following two extremes:
* Prune everything.
* Prune nothing.

Anything in between could be implemented by the caller-supplied KV store; in the simple case, this could be delaying Del operations until e.g. 100 blocks. This would allow us to keep the ProveForRoot etc operations.

This could be implement by using a path -> value KV for this PR, and making pruning optional. The path -> value store (for efficient read ops), would only be for the latest version of the tree. However, the SMT would still need to store valueHash -> value in the nodes MapStore, to allow accessing value for old roots when pruning is disabled. When GetForRoot is called, it would get the leaf by traversing the tree directly, rather than using the path -> value store.

[the remark this answers is not in the document]
Comment (musalbas) [reply deepsubtree.go:36]: Alternatively: this library could support "prune everything" only, but keep storing valueHash -> value in the nodes MapStore (until pruned).

It would then be trivial to implement a "prune nothing" with MapStore scaffolding that simply does nothing when Del is called. GetForRoot and ProveForRoot can then still work. However, those functions would only work for callers that have used a MapStore with this hack.

[the remark this answers is not in the document]
Comment (musalbas) [reply deepsubtree.go:36]: Hmm, actually I just realised that valueHash -> value can't be stored in either proposals, due to the ambiguity issue when pruning. So we need to think of another way to implement pruning while ideally allow for faster Get operations, without forcing people to use BadgerDB or implement their own versioning scheme, if they want to disable pruning.

**Comment by musalbas** [comment]
In light of the above discussion, I'm favouring using the original PR https://github.com/celestiaorg/smt/pull/36 and changing the specification for the following reasons:

* Implementing a path -> value store for faster gets, should actually be the responsibility of the caller, not the SMT, which should only be responsible for the tree itself. Hence, the idea of separating the state commitment library from the storage library. Adding this KV store, would cause them to be coupled.
* Accepting the above, adding a new path -> value store just to implement pruning is wasteful, when we can just make a breaking change to the spec.
* (See discussion above.) Without introducing a breaking change or a lot of extra complexity, it doesn't seem possible to allow people to optionally disable pruning, without forcing them to use BadgerDB (the only well-known key-value-version store), or implement their own versioning.

**Comment by adlerjohn** [comment]
Adding a summary of some nuance discussed on Discord.

More important than breaking the specs (which is unimportant in the grant scheme of things) the real issue is that the current `hash(value) -> value` map, along with the `key || hash(value) -> value` map that replaces it in this PR (and the equivalent in #36), are redundant to the `key -> value` map that will be used in [ADR-40](https://github.com/cosmos/cosmos-sdk/blob/6a5a2de798d4e61302783ae84ecfe2958c7088d5/docs/architecture/adr-040-storage-and-smt-state-commitments.md) to allow for [constant-time leaf reads](https://github.com/celestiaorg/smt/issues/14#issuecomment-762895582). Redundant = space overhead that must be paid by _all_ users of the library, not just users that want historical queries.

In other words, _regardless_ of pruning orphaned nodes, this abstraction must be reworked from the current assumption that a single KV exists for both internal nodes and leaves to the assumption that two KVs exist, one for internal nodes and one for leaves. This second abstract coincidentally aligns exactly with the requirements for constant-time reads, and thus does not require any redundancy.

A suggested alternative would be to store `key -> hash(value)`, then `hash(value) -> value`, adding a level of indirection. But this would then require reads to do two sequential disk accesses instead of one to get a leaf in constant time, which would be an enormous overhead.

Unfortunately, removing redundancy/this mandatory overhead means it becomes impossible to support historical queries, and pruning orphaned nodes must be enabled at all times.

**Comment by roysc** [comment]
This may be a silly question, but what about just removing `Get` value access entirely? The tree could just store its internal structure and value hashes, assuming only the responsibility of proof generation. That would completely avoid redundancy and leave value storage and versioning up to the caller.

**Comment by tzdybal** [comment]
> This may be a silly question, but what about just removing `Get` value access entirely? The tree could just store its internal structure and value hashes, assuming only the responsibility of proof generation. 

This will completely change how this library works. Right now it's a self-contained, general purpose Merkle Tree. Removing `Get` is a fundamental change, meaning that library can't be used "alone", without external store for values.

> That would completely avoid redundancy and leave value storage and versioning up to the caller.

This will only push the problem up to the caller ;)

I agree with @adlerjohn that two KV stores should be used in library (which of course can be implemented as just prefix-buckets in single database).

**Comment by roysc** [comment]
> This will only push the problem up to the caller ;)

Well, that's the idea. The library user can optimize for whatever they need, and the library doesn't have to try to solve for all cases. You could still provide a self-contained tree map that includes value access via one of the methods discussed. Anyway, just a thought.

**Comment by musalbas** [comment]
I agree that given a trade-off between duplicating data storage, remove the Get functionality, and removing the ability to access old state roots, the latter is the least undesirable.

If there's no other options, I'm in favour of having two KV stores, where one of the KV stores is a path -> value store, and dropping the functionality of accessing old state roots.

**Review by musalbas — CHANGES_REQUESTED** [review]
I think we have consensus to change smv to a path -> value store, and drop the functionality of accessing old state roots within this library.

**Comment by marbar3778** [comment]
What is the timeline for the proposed changes. I would like to coordinate an audit of the library soonish.

**Comment by roysc** [comment]
@marbar3778 changes are pushed :) I will tag the owners for re-review.

File: deepsubtree.go [hunk 4]
```diff
@@ -33,7 +33,7 @@ func (dsmst *DeepSparseMerkleSubTree) AddBranch(proof SparseMerkleProof, key []b
 	}
 
 	if valueHash != nil {
-		if err := dsmst.values.Set(valueKey(key, valueHash), value); err != nil {
+		if err := dsmst.values.Set(dsmst.SparseMerkleTree.th.path(key), value); err != nil {
```
Comment (adlerjohn) [inline deepsubtree.go:36]: Correct me if I'm wrong, but embedding is done explicitly to avoid having to use `.SparseMerkleTree`. It seems more idiomatic to actually use embedding as intended here.

```suggestion
		if err := dsmst.values.Set(dsmst.th.path(key), value); err != nil {
```

**Review by adlerjohn — APPROVED** [review]

File: proofs.go [hunk 5]
```diff
@@ -162,7 +159,7 @@ func verifyProofWithUpdates(proof SparseMerkleProof, root []byte, key []byte, va
 		updates = append(updates, update)
 	}
 
-	return bytes.Equal(currentHash, root), updates
+	return bytes.Equal(currentHash, root), updates, memberValueHash
```
Comment (musalbas) [inline proofs.go:162]: What's the purpose of returning the valueHash? I don't see this output being used anywhere, except to check if it's nil in AddBranch.

Comment (roysc) [reply proofs.go:162]: Oversight from when valueHash was used to index value - fixed

**Comment by musalbas** [comment]
The core PR LGMT, however there is some other things that should be do as a result of this PR:

1. Update the example in the readme.
2. Remove the API to get old roots.
3. Update Get so that it reads from the path->value store directly, instead of traversing the tree.

**Review by musalbas — APPROVED** [review]

**Comment by adlerjohn** [comment]
I'll handle these in follow-up PRs so as to unblock @roysc. Good work on this PR!
